### ISSUE TITLE
Throw error if validate() is called during validation instead of queuing it up for next frame

### ISSUE THREAD

**Issue by b005t3r** [BowlerHatLLC/feathers#1430]
In classes like FeathersControl, Scale3Image and many others is a really terrible piece of code which hides validation-loop errors from the user.

```as3
//we were already validating, and something else told us to
//validate. that's bad...
if(this._validationQueue)
{
    //...so we'll just try to do it later
    this._validationQueue.addControl(this, true);
}
return;
```

It's just bad, makes to sense to handle it this way and it really pisses me off when I find something like that after hours of bug hunting. Please, Josh, throw and exception when something like that happens, so people would know their code does something which causes problems, don't make us guess and waste our time.

**Comment by joshtynjala** [comment]
Yeah, that's probably a good place to throw an error instead.

**Comment by joshtynjala** [comment]
This change is a more dangerous than I originally thought. Simply throwing the error without any other changes will break some core Feathers classes that try to validate a component before positioning or resizing it. The layout code may be triggered when a component's Event.RESIZE is dispatched, and that event can be dispatched while the component is in the middle of validating. I considered adding a public getter so that you can check if the component is validating before calling validate(), but that would make things too cumbersome because it would require annoying boilerplate. If I ensure that Event.RESIZE is dispatched only at the end of validation, the error won't be thrown. However, it will affect other parts of Feathers that also relied on the same original code to dispatch the event. 

It's not impossible to resolve all of that, but what should be a simple change is becoming more and more complex as I really dig into it. In the paragraph above, that's simply where I happened to decide to stop and rethink what I was doing. If I had continued making all of the required changes, more unexpected side effects may have been revealed. I can think of more events that are dispatched during validation, and they would probably have similar issues.

I'm pretty sure that I specifically added this behavior so that you don't need to check if a component is already validating before you call validate(). Feathers simply tries to validate the component again the next chance it has. However, if the component isn't invalid during that "next chance", it will simply return without doing anything. It won't be added to the queue again if it doesn't need to be. With all that in mind, this code shouldn't exacerbate any issues where a component is invalidating every frame. That component would already be in the queue regardless. Instead, this is actually a helpful feature because it saves you from writing excessive boilerplate.

**Comment by b005t3r** [comment]
But just to be clear - this can lead to a infinite validation loop, where a child triggers parent's validation and the parent in turn forces children to be revalidated, right? If so, this is probably a significant design flaw and what makes things even worse is the fact that Feathers relies on this behavior internally.

**Comment by joshtynjala** [comment]
You've only reworded what I said above. The child will be queued up because it's already validating. If it's not actually invalid, it won't revalidate. If it is invalid, it wasn't caused by this code, and it would have been in queue regardless.

**Comment by b005t3r** [comment]
OK, but I'm talking about the case where the child is invalidated and when it starts to validate it invalidate its parent which in turn invalidates the child - won't this create an infinite validation loop?

**Comment by joshtynjala** [comment]
If the parent validating causes the child to invalidate, that would have happened even if this particular code didn't exist.

**Comment by joshtynjala** [comment]
Actually, though, I realize that this code may not be necessary. It may be better to simply return and do nothing. As I mentioned, if the component actually needs to validate a second time, it will be added to the queue elsewhere. That makes this code redundant. I think I'll try simply returning to see if that has any unexpected side effects.

**Comment by b005t3r** [comment]
> If the parent validating causes the child to invalidate, that would have happened even if this particular code didn't exist.

But this is the bug I've been talking about from the start and the problem is you're not able to track down such bugs, because Feathers ignores such scenarios :) That's why I was asking for an exception if something like that is discovered.

Please, try to imagine you're coding a complex UI. And it work, but after a couple of months you make changes, update dependencies and you've got an infinite validation loop - how would you even start to track that down? And now the same situation, but with Feathers throwing an exception - it would be merely a minor fix.

**Comment by joshtynjala** [comment]
Please try to understand that I actually understand your situation. You don't need to repeat yourself about why this is a problem. I have had to debug this sort of thing too, and it is frustrating.

I want to find a better location to throw an error or print a warning to the console. This is not the right place. There will be more false positives than actual detections of a real problem. However, I haven't given up on finding somewhere else where I can do that.

> how would you even start to track that down?

I typically track down this type of issue like this:

1) If it's obvious which component is having an issue, I override invalidate() and simply call super.invalidate(). Then, once the app is clearly stuck in the loop, I add a breakpoint. The callstack usually gives very helpful clues about what's causing the invalidation. I might resume the app a few times to see if it hits the same breakpoint for other reasons.

2. If it's not obvious which component is having an issue, I may start by simply pausing the debugger to see what code happens to be running when it's stuck in the loop. Often, it will be in the middle of validating a component, and then I can go back to 1. Worst case scenario, I sometimes add the breakpoint to the main invalidate() in FeathersControl. It usually involves extra noise, but it's still possible to see a pattern in the callstack.

**Comment by b005t3r** [comment]
OK, I just wanted to make sure you're not saying this is unimportant or not worth changing :)

**Comment by b005t3r** [comment]
BTW, pausing the debugger stopped working for me (I have no idea why, I'm using IDEA 14.0.5).

**Comment by joshtynjala** [comment]
It seems to be working for me with IDEA 2016.2 and AIR 22. Hard to say if something changed in the SDK or IDEA, or if something might have gotten corrupted locally, though.

**Comment by joshtynjala** [comment]
I just tentatively called trace() inside invalidate() when _isValidating is true, and it's showing me some potential places where core Feathers components are invalidating when they don't need to. This location seems promising. I don't want to throw an error here because there are legitimate reasons to invalidate during validation, but a one-time console warning per component may be perfect.

**Comment by joshtynjala** [comment]
I found a place in invalidate() where I could throw another error. Similar to the code that you recently pointed out in Scroller, where it would detect an infinite loop during measurement and simply give up, FeathersControl was also giving up silently in invalidate(). If invalidate() is called during validation, the component gets added back to the end of the queue. After 10 tries, it would stop and try again on the next frame in the hopes that the app would continue working and performance wouldn't be hurt too much. Now, it throws an error when this infinite loop is detected.

If I'm not mistaken, this should be the big one that exposes most, or (hopefully) all, of these infinite validation issues immediately. The call stack will expose the exact component that's causing the issue, and the previous function will be the one that is repeatedly calling invalidate() when it shouldn't.

As a bonus, since nothing will wait until the next frame to validate anymore, I was able to simplify some other code too.
